### PR TITLE
[DROOLS-6994] KieContainer.verify() doesn't work with exec-model with…

### PR DESCRIPTION
**Pull request by tkobayas**
…out drools-mvel

Only unit test.

**Ports** 
This is for main. I may backport it to  if it makes sense.

**JIRA**: 

https://issues.redhat.com/browse/DROOLS-6994

<details>
<summary>
How to replicate CI configuration locally?
</summary>

Build Chain tool does "simple" maven build(s), the builds are just Maven commands, but because the repositories relates and depends on each other and any change in API or class method could affect several of those repositories there is a need to use [build-chain tool](https://github.com/kiegroup/github-action-build-chain) to handle cross repository builds and be sure that we always use latest version of the code for each repository.
 
[build-chain tool](https://github.com/kiegroup/github-action-build-chain) is a build tool which can be used on command line locally or in Github Actions workflow(s), in case you need to change multiple repositories and send multiple dependent pull requests related with a change you can easily reproduce the same build by executing it on Github hosted environment or locally in your development environment. See [local execution](https://github.com/kiegroup/github-action-build-chain#local-execution) details to get more information about it.
</details>

<details>
<summary>
How to retest this PR or trigger a specific build:
</summary>

- for <b>pull request checks</b>  
  Please add comment: <b>Jenkins retest this</b>

- for a <b>specific pull request check</b>  
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|optaplanner|kogito-apps|kogito-examples|optaplanner-quickstarts|optaweb-employee-rostering|optaweb-vehicle-routing] tests</b>

- for a <b>full downstream build</b> 
  - for <b>jenkins</b> job: please add comment: <b>Jenkins run fdb</b>
  - for <b>github actions</b> job: add the label `run_fdb`

- <b>a compile downstream build</b> please  add comment: <b>Jenkins run cdb</b>

- <b>a full production downstream build</b> please add comment: <b>Jenkins execute product fdb</b>

- <b>an upstream build</b> please add comment: <b>Jenkins run upstream</b>

- for <b>quarkus branch checks</b>  
  Run checks against Quarkus current used branch  
  Please add comment: <b>Jenkins run quarkus-branch</b>

- for a <b>quarkus branch specific check</b>  
  Run checks against Quarkus current used branch  
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|optaplanner|kogito-apps|kogito-examples|optaplanner-quickstarts|optaweb-employee-rostering|optaweb-vehicle-routing] quarkus-branch</b>

- for <b>quarkus main checks</b>  
  Run checks against Quarkus main branch  
  Please add comment: <b>Jenkins run quarkus-main</b>

- for a <b>specific quarkus main check</b>  
  Run checks against Quarkus main branch  
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|optaplanner|kogito-apps|kogito-examples|optaplanner-quickstarts|optaweb-employee-rostering|optaweb-vehicle-routing] quarkus-branch</b>

- for <b>native checks</b>  
  Run native checks  
  Please add comment: <b>Jenkins run native</b>

- for a <b>specific native check</b>  
  Run native checks 
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|optaplanner|kogito-apps|kogito-examples|optaplanner-quickstarts|optaweb-employee-rostering|optaweb-vehicle-routing] native</b>

- for <b>mandrel checks</b>  
  Run native checks against Mandrel image
  Please add comment: <b>Jenkins run mandrel</b>

- for a <b>specific mandrel check</b>  
  Run native checks against Mandrel image  
  Please add comment: <b>Jenkins (re)run [drools|kogito-runtimes|optaplanner|kogito-apps|kogito-examples|optaplanner-quickstarts|optaweb-employee-rostering|optaweb-vehicle-routing] mandrel</b>
</details>

### CODE REVIEW

**Comment by kie-ci1** [comment]
**(tests) - drools job** `#106` was: **UNSTABLE**
Possible explanation: This should be test failures


**Test results:**
- PASSED: 20818
- FAILED: 1

Those are the test failures: 
<details>
<summary><a href="https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/kogito/job/main/job/pullrequest/job/drools.tests.drools/106/testReport/org.drools.compiler.integrationtests.nomvel/VerifyTest/testVerify/">org.drools.compiler.integrationtests.nomvel.VerifyTest.testVerify</a></summary>
You're trying to compile a Drools asset without mvel. Please add the module org.drools:drools-mvel to your classpath.
</details>

Please look here: https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/kogito/job/main/job/pullrequest/job/drools.tests.drools/106/display/redirect

**Comment by kie-ci3** [comment]
**(tests) - kogito-apps job** `#849` was: **UNSTABLE**
Possible explanation: This should be test failures

Please look here: https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/kogito/job/main/job/pullrequest/job/drools.tests.downstream.kogito-apps/849/display/redirect

**Test results:**
- PASSED: 1960
- FAILED: 1

Those are the test failures: 
<details>
<summary><a href="https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/kogito/job/main/job/pullrequest/job/drools.tests.downstream.kogito-apps/849/testReport/org.kie.kogito.it.jobs/ProcessAsyncIT/Build_projects___testAsync/">Build projects / org.kie.kogito.it.jobs.ProcessAsyncIT.testAsync</a></summary>
java.lang.AssertionError: <br/>1 expectation failed.<br/>JSON path hello doesn't match.<br/>Expected: null<br/>  Actual: Hello Tiago<br/><br/>
</details>

File: drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/CanonicalKieModule.java [hunk 1]
```diff
@@ -897,4 +917,47 @@ public KnowledgeBuilderConfiguration createBuilderConfiguration( KieBaseModel kB
...
+
+    @Override
+    public Optional<KieModuleKieProject> recreateKieProjectForVerify(ClassLoader classLoader) {
+        try {
+            if (!containsRuleAsset()) {
+                LOG.info("This module doesn't contain rule assets." +
+                        " It means this module would correctly work with generated classes, but the verify method produces an empty result." +
+                        " You may assume that verify is successful because the executable model was successfully built beforehand." +
+                        " This may happen when you build a kjar with generateModel=YES option");
+            }
```
Comment (tkobayas) [inline drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/CanonicalKieModule.java:934]: `CanonicalKieModule` may or may not contain original rule assets (drl, etc.). Here, `CanonicalModelKieProject` is recreated by populating files inside `CanonicalKieModule` , so if it doesn't contain original rule assets, `verify` would just successfully finish. Hence, just log the message.

File: drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/CanonicalKieModule.java [hunk 2]
```diff
@@ -897,4 +917,47 @@ public KnowledgeBuilderConfiguration createBuilderConfiguration( KieBaseModel kB
...
+            Class<? extends ProjectType> projectClass = (Class<? extends KieBuilder.ProjectType>) Class.forName("org.drools.model.codegen.ExecutableModelProject");
```
Comment (tkobayas) [inline drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/CanonicalKieModule.java:935]: `Class.forName` is not elegant, but there is no other way because `drools-model-compiler` doesn't depend on `drools-model-codegen`? (And using SPI seems to be overkill)

**Comment by kie-ci3** [comment]
**(tests) - kogito-apps job** `#852` was: **UNSTABLE**
Possible explanation: This should be test failures

Please look here: https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/kogito/job/main/job/pullrequest/job/drools.tests.downstream.kogito-apps/852/display/redirect

**Test results:**
- PASSED: 1960
- FAILED: 1

Those are the test failures: 
<details>
<summary><a href="https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/kogito/job/main/job/pullrequest/job/drools.tests.downstream.kogito-apps/852/testReport/org.kie.kogito.it.jobs/ProcessAsyncIT/Build_projects___testAsync/">Build projects / org.kie.kogito.it.jobs.ProcessAsyncIT.testAsync</a></summary>
java.lang.AssertionError: <br/>1 expectation failed.<br/>JSON path hello doesn't match.<br/>Expected: null<br/>  Actual: Hello Tiago<br/><br/>
</details>

File: drools-model/drools-model-codegen/src/main/java/org/drools/model/codegen/execmodel/generator/ModelGenerator.java [hunk 3]
```diff
@@ -350,9 +352,20 @@ private static List<MethodCallExpr> ruleAttributes(RuleContext context, RuleDesc
...
+    private static void checkDuplicateGroup(Set<Entry<String, AttributeDescr>> excludingDialect, RuleContext context, RuleDescr ruleDescr) {
+        boolean hasAgendaGroup = excludingDialect.stream().anyMatch(entry -> entry.getKey().equals("agenda-group"));
+        boolean hasRuleFlowGroup = excludingDialect.stream().anyMatch(entry -> entry.getKey().equals("ruleflow-group"));
+        if (hasAgendaGroup && hasRuleFlowGroup) {
+            context.addCompilationWarning(new RuleBuildWarning(null, ruleDescr, null,
+                                                               "Both an agenda-group and a ruleflow-group are defined for rule " + ruleDescr.getName() + ". " +
+                                                               "Since version 6.x the two concepts have been unified, the ruleflow-group name will override the agenda-group."));
+        }
+    }
+
```
Comment (tkobayas) [inline drools-model/drools-model-codegen/src/main/java/org/drools/model/codegen/execmodel/generator/ModelGenerator.java:368]: This warning is equivalent to what non-exec-model already has. I added this to exec-model so that we can test `verify` with a warning. (Actually, I cannot find other way to write a rule to cause a warning)

**Comment by kie-ci3** [comment]
**(tests) - drools job** `#876` was: **UNSTABLE**
Possible explanation: This should be test failures

Please look here: https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/kogito/job/main/job/pullrequest/job/drools.tests.drools/876/display/redirect

**Test results:**
- PASSED: 20905
- FAILED: 1

Those are the test failures: 
<details>
<summary><a href="https://eng-jenkins-csb-business-automation.apps.ocp-c1.prod.psi.redhat.com/job/KIE/job/kogito/job/main/job/pullrequest/job/drools.tests.drools/876/testReport/org.kie.pmml.evaluator.core/PMMLRuntimeContextImplTest/getFixedProbabilityMap/">org.kie.pmml.evaluator.core.PMMLRuntimeContextImplTest.getFixedProbabilityMap</a></summary>
bound must be positive
</details>

File: drools-mvel/src/main/java/org/drools/mvel/asm/ClassFieldInspectorImpl.java [hunk 4]
```diff
@@ -507,47 +507,6 @@ public void visitEnd() {
 
     }
 
-    public static class GetterOverloadWarning implements KnowledgeBuilderResult {
```
Comment (mariofusco) [inline drools-mvel/src/main/java/org/drools/mvel/asm/ClassFieldInspectorImpl.java:510]: Is this related with this PR? Or did you simply find that this class is no longer in use?

File: drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/CanonicalKieModule.java [hunk 5]
```diff
@@ -897,4 +917,47 @@ public KnowledgeBuilderConfiguration createBuilderConfiguration( KieBaseModel kB
...
+            KieModuleKieProject kieProject = kprojectSupplier.apply(internalKieModule, classLoader);
+            kieProject.init();
+            return Optional.of(kieProject);
+        } catch (IllegalArgumentException | IllegalAccessException | NoSuchFieldException | SecurityException | ClassNotFoundException e) {
+            LOG.warn("Failed to recreate KieProject. Falls back to default verify", e);
+        }
+        return Optional.empty();
+    }
+
+    private static final Set<String> RULE_ASSET_EXTENSIONS = initRuleAssetExtensions();
```
Comment (mariofusco) [inline drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/CanonicalKieModule.java:946]: Should this probably be moved in the `ResourceType` class? I believe that exposing a Set of known resource types related with rules should be a feature provided by that class.

File: drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/CanonicalKieModule.java (hunk 2 above)
Comment (mariofusco) [reply drools-model/drools-model-compiler/src/main/java/org/drools/modelcompiler/CanonicalKieModule.java:935]: I see the problem and I agree this is the most reasonable solution.

**Review by mariofusco — APPROVED** [review]

File: drools-mvel/src/main/java/org/drools/mvel/asm/ClassFieldInspectorImpl.java (hunk 4 above)
Comment (tkobayas) [reply drools-mvel/src/main/java/org/drools/mvel/asm/ClassFieldInspectorImpl.java:510]: @mariofusco Yes, I just found this class is no longer used. I found this while trying to write a test case rule to produce Warning :)

**Comment by mariofusco** [comment]
Replaced by https://github.com/kiegroup/drools/pull/4884
